src/GraphQLParser/Visitors/SDLWriter.cs

```diff
@@ -0,0 +1,765 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using GraphQLParser.AST;
+
+namespace GraphQLParser.Visitors
+{
+    /// <summary>
+    /// Prints AST into the provided <see cref="TextWriter"/> as a SDL document.
+    /// </summary>
+    /// <typeparam name="TContext">Type of the context object passed into all VisitXXX methods.</typeparam>
+    public class SDLWriter<TContext> : DefaultNodeVisitor<TContext>
+        where TContext : IWriteContext
+    {
+        /// <inheritdoc/>
+        public override async ValueTask VisitDocument(GraphQLDocument document, TContext context)
+        {
+            if (document.Definitions?.Count > 0)
+            {
+                for (int i = 0; i < document.Definitions.Count; ++i)
+                {
+                    await Visit(document.Definitions[i], context);
+
+                    if (i < document.Definitions.Count - 1)
+                    {
+                        await context.Writer.WriteLineAsync();
+                    }
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public override async ValueTask VisitComment(GraphQLComment comment, TContext context)
+        {
+            int level = GetLevel(context);
+
+            bool needStartNewLine = true;
+            int length = comment.Text.Span.Length;
+            for (int i = 0; i < length; ++i)
+            {
+                if (needStartNewLine)
+                {
+                    await WriteIndent(context, level);
+                    await context.Writer.WriteAsync('#');
+                    needStartNewLine = false;
+                }
+
+                char current = comment.Text.Span[i];
+                switch (current)
+                {
+                    case '\r':
+                        break;
+
+                    case '\n':
+                        await context.Writer.WriteLineAsync();
+                        needStartNewLine = true;
+                        break;
+
+                    default:
+                        await context.Writer.WriteAsync(current);
```

Comment:
It would be nice to use the cancellationtoken within this method... but it's an optimization that can be done later.

Reply:
Done, all writes into `TextWriter` now are handled via extension methods respecting token from context.